### PR TITLE
User crud

### PR DESCRIPTION
**Pull request by ytrufchev**

### CODE REVIEW

File: src/main/java/com/lms/learning_management_system/entities/service/UserService.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    public UserDTO createUser(UserDTO userDTO) {
+        UserEntity userEntity = convertToEntity(userDTO);
+        userEntity = userRepository.save(userEntity);
+        return convertToDTO(userEntity);
+    }
+
+    public UserDTO updateUser(UUID id, UserDTO userDTO) {
+        UserEntity userEntity = userRepository.findById(id)
+                .orElseThrow(() -> new UserNotFoundException("User with id " + id + " not found"));
```
Comment (ignatIgnatov) [inline src/main/java/com/lms/learning_management_system/entities/service/UserService.java]: You can export it in a method

File: src/main/java/com/lms/learning_management_system/entities/service/UserService.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        return userRepository.findById(id)
+                .map(this::convertToDTO)
+                .orElseThrow(() -> new UserNotFoundException("User with id " + id + " not found"));
+    }
+
+    public UserDTO createUser(UserDTO userDTO) {
+        UserEntity userEntity = convertToEntity(userDTO);
+        userEntity = userRepository.save(userEntity);
+        return convertToDTO(userEntity);
+    }
```
Comment (ignatIgnatov) [inline src/main/java/com/lms/learning_management_system/entities/service/UserService.java]: What happens when the user exists in db

File: src/main/java/com/lms/learning_management_system/entities/service/UserService.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    public UserDTO updateUser(UUID id, UserDTO userDTO) {
+        UserEntity userEntity = userRepository.findById(id)
+                .orElseThrow(() -> new UserNotFoundException("User with id " + id + " not found"));
+
+        userEntity.setFirstName(userDTO.getFirstName());
+        userEntity.setLastName(userDTO.getLastName());
+        userEntity.setEmail(userDTO.getEmail());
+        // Assuming role mapping is handled elsewhere
+        // userEntity.setRole(userDTO.getRole());
```
Comment (ignatIgnatov) [inline src/main/java/com/lms/learning_management_system/entities/service/UserService.java]: Maybe you can use here converToEntity mapper too

File: src/main/java/com/lms/learning_management_system/entities/service/UserService.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    private UserEntity convertToEntity(UserDTO userDTO) {
+        UserEntity userEntity = new UserEntity();
+        RoleEntity roleEntity = new RoleEntity();
+        userEntity.setFirstName(userDTO.getFirstName());
+        userEntity.setLastName(userDTO.getLastName());
+        userEntity.setEmail(userDTO.getEmail());
+        roleEntity.setRole(userDTO.getRole());
+        userEntity.setRole(roleEntity);
+        return userEntity;
+    }
```
Comment (ignatIgnatov) [inline src/main/java/com/lms/learning_management_system/entities/service/UserService.java]: This method should be able to be done by the mapper

**Review by ignatIgnatov — COMMENTED** [review]
I left a few small comments

File: src/main/java/com/lms/learning_management_system/entities/service/UserService.java [hunk 5]
```diff
@@ -34,7 +34,12 @@ public UserDTO getUserById(UUID id) {
     }
 
     public UserDTO createUser(UserDTO userDTO) {
-        UserEntity userEntity = convertToEntity(userDTO);
+        Optional<UserEntity> existingUser = userRepository.findByEmail(userDTO.getEmail());
```
Comment (ignatIgnatov) [inline src/main/java/com/lms/learning_management_system/entities/service/UserService.java:37]: You do a similar operation several times in UserService. For cleaner code, you can make a method and reuse it. Think about it

File: src/main/java/com/lms/learning_management_system/entities/service/UserService.java [hunk 6]
```diff
@@ -43,12 +48,7 @@ public UserDTO updateUser(UUID id, UserDTO userDTO) {
         UserEntity userEntity = userRepository.findById(id)
                 .orElseThrow(() -> new UserNotFoundException("User with id " + id + " not found"));
```
Comment (ignatIgnatov) [inline src/main/java/com/lms/learning_management_system/entities/service/UserService.java]: Same comment as above...

**Review by ignatIgnatov — COMMENTED** [review]
Looks better! Some litle comments about similar repeating operations.
The tests look good. But think about it for the future: we can make a Helper Class in which we export all the reusable methods of the final fields. This way the code will look cleaner and neater

File: src/main/java/com/lms/learning_management_system/entities/service/UserService.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+@Service
+public class UserService {
+
+    private final UserRepository userRepository;
+    String userWithId = "User with Id";
+    String notFound = "Not found";
+    @Autowired
+    public UserService(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
```
Comment (ignatIgnatov) [inline src/main/java/com/lms/learning_management_system/entities/service/UserService.java]: The string must be private static final.
If you use @RequireArgsConstructor from Lombok for UserService class, then you will save 2-3 rows of code, and the code will be cleaner

File: src/main/java/com/lms/learning_management_system/entities/service/UserService.java (hunk 3 above)
Comment (ytrufchev) [reply src/main/java/com/lms/learning_management_system/entities/service/UserService.java]: fixed

File: src/main/java/com/lms/learning_management_system/entities/service/UserService.java (hunk 7 above)
Comment (ytrufchev) [reply src/main/java/com/lms/learning_management_system/entities/service/UserService.java]: done

File: src/main/java/com/lms/learning_management_system/entities/service/UserService.java (hunk 5 above)
Comment (ytrufchev) [reply src/main/java/com/lms/learning_management_system/entities/service/UserService.java:37]: I am not exactly sure how to, i use different stream operation after the findBy and i don't know how to implement that

Comment (ignatIgnatov) [reply src/main/java/com/lms/learning_management_system/entities/service/UserService.java:37]: ok leave it at that

File: src/main/java/com/lms/learning_management_system/controller/UserController.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Autowired
+    public UserController(UserService userService) {
+        this.userService = userService;
+    }
+
+    @PostMapping
+    public UserDTO createUser(@RequestBody UserDTO userDTO) {
+        return userService.createUser(userDTO);
+    }
```
Comment (ignatIgnatov) [inline src/main/java/com/lms/learning_management_system/controller/UserController.java:29]: You can catch up @ResponseStatus(HttpStatus.CREATED)

Comment (ytrufchev) [reply src/main/java/com/lms/learning_management_system/controller/UserController.java:29]: done

**Comment by ShirinAI** [comment]
It is okay, please merge

**Review by ShirinAI — APPROVED** [review]
Approved

**Pull request merged**